types/react-instantsearch-native/index.d.ts

```diff
@@ -60,6 +55,18 @@ export interface InstantSearchProps {
         props: any;
     };
 }
+
+export interface UsingSearchClient extends InstantSearchBaseProps {
+    searchClient: any;
+}
+
+export interface UsingManualInfo extends InstantSearchBaseProps  {
+    apiKey: string;
+    appId: string;
+    algoliaClient?: any;
+}
+
+export type InstantSearchProps = UsingSearchClient | UsingManualInfo;
```

Comment:
I could be missing something obvious, but it looks like this makes accessing `apiKey` on an object of type `InstantSearchProps` an error.  Won't that break existing consumers?

Reply:
Does it mean you use this prop type? 
In the case, we may use UsingManualInfoProps or type confinement. This PR is stricter than ever, so someone may meet breaking change but it's reasonable as I thought it.

Reply:
Or I'm sorry if I misunderstand what you mean. If you mentioning when initializing IntantSearch components, we need either props
```
 {
  apiKey: string;
  appId: string;
  algoliaClient?: any;
}
``` 
or
```
 {
   searchClient: any;
 }
```
as required. So you need apiKey with appId parameters, too. Could I answer it?

Reply:
I agree with @tkow

Reply:
Another option is to add `searchClient: undefined` to the `UsingManualInfo` definition, and vice versa.

Reply:
Sorry, I think I was unclear.  Suppose you had 
```ts
type T = { a: number } | { b: number };
let t: T;
```
Then `t` would have either a property called `a` or a property called `b`, which I think is what you're going for.  However, as written, both `t.a` and `t.b` are errors because the union type has neither of those properties.

Reply:
@amcasey yes I've run into just this problem in my typescript projects.  You are correct!

Supposing your example, you could not do this:

```ts
let t: T = ...
console.log(t.a)  // ERROR!
```

You have to do this:

```ts
let t: T = ...
if ('a' in t) {
  console.log(t.a)
}
```

However, if you define the types like this:
```ts
type T = { a: number, b?: undefined } | { a?: undefined, b: number };
```

then you can do this:
```ts
let t: T = ...
if (t.a) {
  console.log(t.a)
}
```

Reply:
@gburgett 
You describe clearly what I want to tell @amcasey. And as I see it, InstantSearchProps are useless type about exporting your project in most case, because, when you want to use this props' parameter in your callback args or your components' props,   you can achieve  UsingSearchClientProps or UsingManualInfoProps instead of InstantSearchProps in your components. It's more preferable to before because typescript will not know which props used when initialize in these scope. It depends on developers's choice. So, I don't have any idea about case we need UsingSearchClientProps type except for you want to pass the props from upper parent keeping ambiguous initialize props. In this context, we don't need  https://github.com/DefinitelyTyped/DefinitelyTyped/pull/38125#discussion_r321316445 change as I thought it.